shared_model/interfaces/iroha_internal/transaction_batch.hpp

```diff
@@ -0,0 +1,94 @@
+       * @tparam TransactionValidator validates every single transaction
+       * @tparam OrderValidator validates order of transactions
+       * @param transactions collection of transactions, should be from the same
+       * batch
+       * @param validator transactions collection validator with provided
+       * transaction validator and order validator
+       * @return valid batch of transactions
+       */
+      template <typename TransactionValidator, typename OrderValidator>
+      static iroha::expected::Result<TransactionBatch, std::string>
+      createTransactionBatch(const types::SharedTxsCollectionType &transactions,
+                             const validation::TransactionsCollectionValidator<
+                                 TransactionValidator,
+                                 OrderValidator> &validator);
+
+      /**
+       * Creates transaction batch from single transaction
+       * @tparam TransactionValidator validates every single transaction
+       * @param transaction is transaction being validated and used to create
+       * batch
+       * @param transaction_validator transaction validation logic
+       * @return batch with single transaction
+       * @note transactions in such batches may not have batch meta information
+       */
+      template <typename TransactionValidator>
+      static iroha::expected::Result<TransactionBatch, std::string>
+      createTransactionBatch(std::shared_ptr<Transaction> transaction,
+                             const TransactionValidator &transaction_validator =
+                                 TransactionValidator());
+
+      /**
+       * Get transactions list
+       * @return list of transactions from the batch
+       */
+      const types::SharedTxsCollectionType &transactions() const;
+
+      /**
+       * Get the concatenation of reduced hashes as a single hash
+       * @param reduced_hashes collection of reduced hashes
+       * @return concatenated reduced hashes
+       */
+      const types::HashType &reducedHash() const;
+
+      /**
+       * Checks if every transaction has quorum signatures
+       * @return true if every transaction has quorum signatures, false
+       * otherwise
+       */
+      bool hasAllSignatures() const;
+
+      /**
+       * Get the concatenation of reduced hashes as a single hash
+       * That kind of hash does not respect batch type
+       * @param reduced_hashes
+       * @return concatenated reduced hashes
+       */
+      static types::HashType calculateReducedBatchHash(
+          const boost::any_range<types::HashType, boost::forward_traversal_tag>
+              &reduced_hashes);
+
+     private:
+      explicit TransactionBatch(
+          const types::SharedTxsCollectionType &transactions)
+          : transactions_(transactions) {}
+
+      types::SharedTxsCollectionType transactions_;
+
+      mutable boost::optional<types::HashType> reduced_hash_;
```

Comment:
Explain why this field is mutable. Seems class is immutable and therefore reduced hash is a constant.

Reply:
`reduced_hash_` contains value of hash if it is calculated and none otherwise, since that field is updated in const method `reducedHash()` when it is invoked very first time. This is exactly the case when mutable is usually used (besides lambdas) http://alenacpp.blogspot.com/2005/10/mutable-constcast.html